llarp/handlers/tun.cpp

```diff
@@ -740,140 +693,55 @@ namespace llarp
-          if (impl && impl->parent && impl->parent->tick)
-            impl->parent->tick(impl->parent);
-          llarp::LogTrace("TunEndpoint ticker() end");
-        };
-        if (not loop->add_ticker(ticker))
-        {
-          llarp::LogError(Name(), " failed to add vpn to event loop");
-          if (vpnif->injected)
-            vpnif->injected(vpnif, false);
-          return false;
-        }
-      }
-      else
-      {
-        if (!llarp_ev_add_tun(loop.get(), tunif.get()))
-        {
-          llarp::LogError(
-              Name(), " failed to set up tun interface: ", tunif->ifaddr, " on ", tunif->ifname);
-          return false;
-        }
-      }
-      const char* ifname;
-      const char* ifaddr;
-      unsigned char netmask;
-      if (tunif)
-      {
-        ifname = tunif->ifname;
-        ifaddr = tunif->ifaddr;
-        netmask = tunif->netmask;
-      }
-      else
+      const service::Address ourAddr = m_Identity.pub.Addr();
+
+      if (not MapAddress(ourAddr, GetIfAddr(), false))
       {
-        ifname = vpn.info.ifname;
-        ifaddr = vpn.info.ifaddr;
-        netmask = vpn.info.netmask;
+        return false;
       }
-      m_IfName = ifname;
-      if (ip.FromString(ifaddr))
+
+      vpn::InterfaceInfo info;
+      info.addrs.emplace(m_OurRange);
+      info.ifname = m_IfName;
+      info.dnsaddr.FromString(m_LocalResolverAddr.toHost());
+
+      LogInfo(Name(), " setting up network...");
+
+      try
       {
-        m_OurIP = net::ExpandV4(ip);
-        m_OurRange.netmask_bits = netmask_ipv6_bits(netmask + 96);
+        m_NetIf = Router()->GetVPNPlatform()->ObtainInterface(std::move(info));
       }
-      else if (m_OurIP.FromString(ifaddr))
+      catch (std::exception& ex)
       {
-        m_OurRange.netmask_bits = netmask_ipv6_bits(netmask);
-        m_UseV6 = true;
+        LogError(Name(), " failed to set up network interface: ", ex.what());
       }
-      else
+      if (not m_NetIf)
       {
-        LogError(Name(), " invalid interface address given, ifaddr=", ifaddr);
-        if (vpnif && vpnif->injected)
-          vpnif->injected(vpnif, false);
+        LogError(Name(), " failed to obtain network interface");
         return false;
       }
+      m_IfName = m_NetIf->IfName();
+      LogInfo(Name(), " got network interface ", m_IfName);
 
-      m_NextIP = m_OurIP;
-      m_OurRange.addr = m_OurIP;
-      m_MaxIP = m_OurRange.HighestAddr();
-      llarp::LogInfo(Name(), " set ", ifname, " to have address ", m_OurIP);
-      llarp::LogInfo(Name(), " allocated up to ", m_MaxIP, " on range ", m_OurRange);
-
-      const service::Address ourAddr = m_Identity.pub.Addr();
-
-      if (not MapAddress(ourAddr, GetIfAddr(), false))
+      auto netloop = Router()->netloop();
+      if (not netloop->add_network_interface(
+              m_NetIf, [&](net::IPPacket pkt) { HandleGotUserPacket(std::move(pkt)); }))
       {
+        LogError(Name(), " failed to add network interface");
```

Comment:
error string feels a bit ambiguous.  in a sea of other error messages, it may be unclear what "add network interface" means in this context.

Reply:
any suggestions on how to improve this?

Reply:
maybe something like "failed to add network interface object to net loop" or similar